DOCS-2095: First draft of controls package docs update (#2739)
viamrobotics/docs: docs/internals/controls-package.md

weight: 10
type: "docs"
description: "The Control Package implements feedback control on an endpoint."
# SME: Nicolas Menard, Rand, Martha
---

[Encoded motors](/components/motor/gpio/encoded-motor/) and [sensor controlled bases](/components/base/sensor-controlled/) use a control loop that is implemented by `viam-server`.

> You can configure the `control_parameters` attribute forn both components to adjust the control loop.

Author: ```suggestion
You can configure the `control_parameters` attribute for both components to adjust the control loop.
```

| `NeedsAutoTuning` | bool | True when the loop needs to be auto-tuned. This will be set to true automatically if all PID values are 0. <br> Default: false |
| `LoopFrequency` | float64 | The frequency at which the control loop should run. <br> Default: 50 Hz |
| `ControllableType` | string | The type of component the control loop will be set up for, currently a base or motor. <br> Default: `"motor_name"` |

```go
type Options struct {
  PositionControlUsingTrapz bool
  SensorFeedback2DVelocityControl bool
  DerivativeType string
  UseCustomConfig bool
  CompleteCustomConfig Config
  NeedsAutoTuning bool
  LoopFrequency float64
  ControllableType string
}
```

The built in control loop setup is only structured to work with an encoded motor or a sensor controlled base. If you wish to use a different setup, you can utilize the options by setting `UseCustomConfig` to `true` and `CompleteCustomConfig` to your custom control loop config of type `control.Config`. The [Control Blocks](/internals/controls-package/#controlblocks) section details the different options for control blocks and how to create a `control.Config`.

### BlockConfig

> `BlockConfig` is struct of the configuration for an individual control block. You will have to build individual BlockConfigs if you utilize the `UseCustomConfig` option. Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#controlblocks) section.

Author: this and the link on line 265 also aren't passing checks... but I think that probably makes sense because there's a space in the `Control Blocks` title... how do I account for that in the hash?

## Introduction

> The control package implements feedback control on an endpoint (plant).

Reviewer: I don't know why this said plant haha

The following functions are available for creating and using a control loop:

<!-- prettier-ignore -->
Method Name | Description
----------- | -----------
[`SetupPIDControlConfig`](/internals/controls-package/#setuppidcontrolconfig) | Creates a [PIDLoop](/internals/controls-package/#pidloop) object that contains all the necessary attributes to run a control loop based on the specified [Options](/internals/controls-package/#options).
[`TunePIDLoop`](/internals/controls-package/#tunepidloop) | Automatically tunes the system and logs the calculated PID values for the loop.
[`StartControlLoop`](/internals/controls-package/#startcontrolloop) | Starts the control loop in a background thread.
[`CreateConstantBlock`](/internals/controls-package/#createconstantblock) | Creates a control block of type `constant`, all control loops need at least one constant block representing the set point.
[`UpdateConstantBlock`](/internals/controls-package/#updateconstantblock) | Updates the value of a constant block to the new set point.

> [`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`, control loops, such as for motors, that control position need a trapezoidal velocity profile block.

Author: ```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, encoded motors), need a trapezoidal velocity profile block.
```
does this make more sense

> [`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`, control loops, such as for motors, that control position need a trapezoidal velocity profile block.

Reviewer: yes I think maybe add a "for"
```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, control loops for encoded motors), need a trapezoidal velocity profile block.
```

[`UpdateTrapzBlock`](/internals/controls-package/#updatetrapzblock) | Updates the attributes of a trapezoidal velocity profile block to the new desired max velocity.

### SetupPIDControlConfig

Creates a [PIDLoop](/internals/controls-package/#pidloop) object, which contains all the attributes related to a control loop that a controlled component would need, most importantly, the control config.

**Parameters:**

- `pidVals` [([]PIDConfig)](/internals/controls-package/#pidconfig): The P, I, and D values for the control loop, if all are zero the loop will be auto-tuned and log the calculated PID values.
- `componentName` [(string)](https://pkg.go.dev/builtin#string): The name of the component that the created PID loop will control.

> - `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.

Author: yes starting line 237

> - `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.

Reviewer: Sorry I missed the that the links were already there

> - `c` [(Controllable)](/internals/controls-package/#controllable): An interface that contains the necessary functions to move the controlled component.

Author: yes starting line 221

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A Context carries a deadline, a cancellation signal, and other values across API boundaries.
- `name` [(string)](https://pkg.go.dev/builtin#string): The desired name of the trapezoidal block.
- `maxVel` [(float64)](https://pkg.go.dev/builtin#float64): The max velocity for the controlled component to move at.
- `dependsOn` [([]string)](https://pkg.go.dev/builtin#string): A slice strings containing the names of the other control blocks that the new trapezoidal block depends on. Usually the set point and the end point.

**Returns:**

[(BlockConfig)](/internals/controls-package/#blockconfig): The config for the newly created block.

```go
trapzBlock := control.CreateTrapzBlock(context.Background(), "set_point", 10.0, []string{"set_point", "endpoint"})
```

### UpdateTrapzBlock

Creates a new control block of type `trapezoidalVelocityProfile`, and then updates the control loop to use this new block.

**Parameters:**

- `ctx` [(Context)](https://pkg.go.dev/context): A context carries a deadline, a cancellation signal, and other values across API boundaries.

> - `name` [(string)](https://pkg.go.dev/builtin#string): The desired name of the trapezoidal block.

Author: it's the name for the block that's being created by the function. so it could be anything. to replace an old block the name must be the same but if creating a new one completely it can be named anything

> - `name` [(string)](https://pkg.go.dev/builtin#string): The desired name of the trapezoidal block.

Reviewer: Ok but whatever I specify here will be used and not changed by other code? Then we want to remove "desired" because it can imply that you're only specifying a preference

> - `name` [(string)](https://pkg.go.dev/builtin#string): The desired name of the trapezoidal block.

Author: yeah we don't want it to be changed it should be the same every time... that make sense

- `maxVel` [(float64)](https://pkg.go.dev/builtin#float64): The max velocity for the controlled component to move at.
- `dependsOn` [([]string)](https://pkg.go.dev/builtin#string): A slice strings containing the names of the other control blocks that the new trapezoidal block depends on. Usually the set point and the end point.
- `loop` (`*Loop`): The control loop that needs to be updated with the newly created trapezoidal block.

**Returns:**

- [(error)](https://pkg.go.dev/builtin#error): An error, if one occurred.

```go
err := control.UpdateTrapzBlock(context.Background(), "set_point", 10.0, []string{"set_point", "endpoint"}, loop)
```

### PIDLoop

`PIDLoop` is a struct containting all the attributes used for setting up a PID control loop. [`SetupPIDControlConfig`](/internals/controls-package/#setuppidcontrolconfig) will create this object for you.

```go
type PIDLoop struct {
  BlockNames   map[string][]string
  PIDVals      []PIDConfig
  ControlConf  Config
  ControlLoop  *Loop
  Options      Options
  Controllable Controllable
}
```

> ### PIDConfig

Author: you don't really need it but if you want to really customize things and do it manually you would need it

  Type string
  P    float64
  I    float64
  D    float64
}
```

### Controllable

`Controllable` is an interface that contains the two functions that any component needs to be controlled by a control loop. For any components other than an encoded motor and a sensor controlled base, these functions must be implemented on the component.

- `State()` gets the current state of the endpoint (controlled component) and passes that information on to the next iteration of the control loop
- `SetState()` takes the information from the latest iteration of the control loop and sets the state of the endpoint (controlled component) to the calculated value.

For example, in an encoded motor, `State()` measures the current position of the motor, so that the next iteration of the control loop knows how far it is from the goal position, or how much error remains. Then the control loop calculates what the next power percentage should be in order to get the motor to its goal position and velocity, and `SetState()` sets that power on the motor.

```go
type Controllable interface {
  SetState(ctx context.Context, state []*Signal) error
  State(ctx context.Context) ([]float64, error)
}
```

> ### Options

Author: yep this is linked above where options is first mentioned

> ### Options

Reviewer: Yep sorry i missed that

| `NeedsAutoTuning` | bool | True when the loop needs to be auto-tuned. This will be set to true automatically if all PID values are 0. <br> Default: false |
| `LoopFrequency` | float64 | The frequency at which the control loop should run. <br> Default: 50 Hz |
| `ControllableType` | string | The type of component the control loop will be set up for, currently a base or motor. <br> Default: `"motor_name"` |

```go
type Options struct {
  PositionControlUsingTrapz bool
  SensorFeedback2DVelocityControl bool
  DerivativeType string
  UseCustomConfig bool
  CompleteCustomConfig Config
  NeedsAutoTuning bool
  LoopFrequency float64
  ControllableType string
}
```

The built in control loop setup is only structured to work with an encoded motor or a sensor controlled base. If you wish to use a different setup, you can utilize the options by setting `UseCustomConfig` to `true` and `CompleteCustomConfig` to your custom control loop config of type `control.Config`. The [Control Blocks](/internals/controls-package/#controlblocks) section details the different options for control blocks and how to create a `control.Config`.

### BlockConfig

> `BlockConfig` is struct of the configuration for an individual control block. You will have to build individual BlockConfigs if you utilize the `UseCustomConfig` option. Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#controlblocks) section.

Reviewer: You can see it here but essentially it's a dash for the space :) https://docs-test.viam.dev/2739/internals/controls-package/#control-blocks

```suggestion
`BlockConfig` is a struct for the configuration of an individual control block.
You have to build individual BlockConfigs if you use the `UseCustomConfig` option.
Each block type requires different attributes, which are outlined in the [Control Blocks](/internals/controls-package/#control-blocks) section.
```

The following functions are available for creating and using a control loop:

<!-- prettier-ignore -->
Method Name | Description
----------- | -----------
[`SetupPIDControlConfig`](/internals/controls-package/#setuppidcontrolconfig) | Creates a [PIDLoop](/internals/controls-package/#pidloop) object that contains all the necessary attributes to run a control loop based on the specified [Options](/internals/controls-package/#options).
[`TunePIDLoop`](/internals/controls-package/#tunepidloop) | Automatically tunes the system and logs the calculated PID values for the loop.
[`StartControlLoop`](/internals/controls-package/#startcontrolloop) | Starts the control loop in a background thread.
[`CreateConstantBlock`](/internals/controls-package/#createconstantblock) | Creates a control block of type `constant`, all control loops need at least one constant block representing the set point.
[`UpdateConstantBlock`](/internals/controls-package/#updateconstantblock) | Updates the value of a constant block to the new set point.

> [`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`, control loops, such as for motors, that control position need a trapezoidal velocity profile block.

Author: ```suggestion
[`CreateTrapzBlock`](/internals/controls-package/#createtrapzblock) | Creates a control block of type `trapezoidalVelocityProfile`. Control loops that control position (for example, encoded motors), need a trapezoidal velocity profile block.
```
does this make more sense

[`UpdateTrapzBlock`](/internals/controls-package/#updatetrapzblock) | Updates the attributes of a trapezoidal velocity profile block to the new desired max velocity.

### SetupPIDControlConfig

Creates a [PIDLoop](/internals/controls-package/#pidloop) object, which contains all the attributes related to a control loop that a controlled component needs, including, most importantly, the control config.

**Parameters:**

- `pidVals` [([]PIDConfig)](/internals/controls-package/#pidconfig): The P, I, and D values for the control loop, if all are zero the loop will auto-tune and log the calculated PID values.
- `componentName` [(string)](https://pkg.go.dev/builtin#string): The name of the component that the PID loop controls.

> - `options` [(Options)](/internals/controls-package/#options): All the desired optional parameters to customize the control loop.

Reviewer: ```suggestion
- `options` [(Options)](/internals/controls-package/#options): All the desired [optional parameters](#options) to customize the control loop.
```

Author: is it necessary to link twice in the same statement? if this is the norm that's fine I don't feel strongly about it but I'm just wondering if we need the type "Options" and the phrase "optional parameters" to link to the same spot

## Introduction

> The easiest way to utilize the controls package is by configuring the `control_parameters` attribute for an [encoded motor](/docs/components/motor/gpio/encoded-motor/) and/or a [sensor-controlled base](/docs/components/base/sensor-controlled/). However, if you want to change or customize the control loops on these components, or you want to add a control loop to a different component, you can use the controls package to [build your own PID control loop](/internals/controls-package/#creating-and-using-a-pid-control-loop).

Author: not the most info here, let me know if you have any suggestions on expanding on this

### Endpoint

The Endpoint is a special type of block that is used to represent a plant.
For now, only DC motors with an encoder, and a base with a linear and angular movement sensor are supported as an endpoint in the control package.

> The _motor_name_ attribute should be changed to _base_name_ when using a base.

Reviewer: We try to speak to the user as much as possible because otherwise they may think we will change it. This makes it clear they need to perform the action
```suggestion
You should change the _motor_name_ attribute to _base_name_ when using a base.
```

The following implementation records the step response of the plant and uses the relay method to estimate the ultimate gain (_Ku_) and oscillation period (_Tu_) of the plant.

Several methods to calculate Kp, Ki and Kd are implemented:

> - ziegerNicholsPI

Reviewer: well these seem like very random names. Do we want to explain any of this further?

Author: they are tuning different tuning methods for a PID loop. I don't think we need to add more info as it is very technical and not really viam related. maybe a note that you can google them if you want more info but idk if that's necessary

Reviewer: Ok let's leave as is
